src/boot/mexpr.ml

```diff
@@ -393,7 +393,20 @@ let rec debruijn env t =
   let rec find fi env n x =
     (match env with
      | VarTm(y)::ee -> if y =. x then n else find fi ee (n+1) x
-     | [] -> raise_error fi ("Unknown variable '" ^ Ustring.to_utf8 x ^ "'"))
+     | [] -> raise_error fi ("Unknown variable '" ^ Ustring.to_utf8 x ^ "'")) in
+  let rec dbPat env = function
+    | PatNamed(fi,x) -> (VarTm(x)::env, PatNamed(fi,x))
+    | PatTuple(fi,ps) -> (* NOTE: this causes patterns to introduce names right-to-left, which will cause errors if a later pattern binds a name that is seen as a constructor in a pattern to the left *)
```

Comment:
Mostly yes, but it will at least be solved by #31.

It does still leave shadowing for normal names though, so 

```
match e with (a, a) then a else ...
```

will use the left value, since it shadows the right. This may or may not be expected.

Reply:
My feeling is that shadowing in the same pattern is also something that a typechecker should prevent.